File: appMY/src/main/java/com/giosis/util/qdrive/message/MessageDetailAdapter.java
package com.giosis.util.qdrive.message;

import android.content.Context;
import android.os.Build;
import android.text.Html;
import android.util.Log;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.giosis.util.qdrive.international.R;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;

public class MessageDetailAdapter extends BaseAdapter {

    Context mContext;

    RelativeLayout layout_message_detail;

    LinearLayout layout_message_detail_receive_message;
    TextView text_message_detail_receiver_id;
    TextView text_message_detail_receive_message;
    TextView text_message_detail_receive_date;

    RelativeLayout layout_message_detail_send_message;
    TextView text_message_detail_send_date;
    TextView text_message_detail_send_message;

    ArrayList<MessageDetailResult.MessageDetailList> messageDetailListArrayList;
    String calledFragment;


    public MessageDetailAdapter(Context context, ArrayList<MessageDetailResult.MessageDetailList> item, String called) {

        this.mContext = context;
        this.messageDetailListArrayList = item;
        this.calledFragment = called;
    }

    @Override
    public int getCount() {
        if (messageDetailListArrayList != null && 0 < messageDetailListArrayList.size()) {
            return messageDetailListArrayList.size();
        }

        return 0;
    }

    @Override
    public Object getItem(int position) {
        return messageDetailListArrayList.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        View view;

        if (convertView == null) {

            LayoutInflater inflater = (LayoutInflater) mContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            view = inflater.inflate(R.layout.message_detail_item, null);
        } else {

            view = convertView;
        }

        layout_message_detail = view.findViewById(R.id.layout_message_detail);

        layout_message_detail_receive_message = view.findViewById(R.id.layout_message_detail_receive_message);
        text_message_detail_receiver_id = view.findViewById(R.id.text_message_detail_receiver_id);
        text_message_detail_receive_message = view.findViewById(R.id.text_message_detail_receive_message);
        text_message_detail_receive_date = view.findViewById(R.id.text_message_detail_receive_date);

        layout_message_detail_send_message = view.findViewById(R.id.layout_message_detail_send_message);
        text_message_detail_send_date = view.findViewById(R.id.text_message_detail_send_date);
        text_message_detail_send_message = view.findViewById(R.id.text_message_detail_send_message);


        MessageDetailResult.MessageDetailList item = messageDetailListArrayList.get(position);


        if (item.getAlign().equalsIgnoreCase("LEFT")) {
            /* from customer, admin > to driver */
            layout_message_detail_receive_message.setVisibility(View.VISIBLE);
            layout_message_detail_send_message.setVisibility(View.GONE);

            if (item.getSender_id() == null || item.getSender_id().equals("")) {

                text_message_detail_receiver_id.setText("null");
            } else {

                text_message_detail_receiver_id.setText(item.getSender_id());
            }
            text_message_detail_receive_date.setText(item.getSend_date());

            // Customer(Admin) ID   visible/gone
            if (position == 0) {

                text_message_detail_receiver_id.setVisibility(View.VISIBLE);
            } else {  // 1 이상일 때, 나의 이전  ALIGN  비교

                String prev_align = messageDetailListArrayList.get(position - 1).getAlign();
                String this_align = messageDetailListArrayList.get(position).getAlign();

                if (!this_align.equals(prev_align)) {   // right > left

                    text_message_detail_receiver_id.setVisibility(View.VISIBLE);
                } else {

                    String prev_date_string = messageDetailListArrayList.get(position - 1).getSend_date();
                    String this_date_string = messageDetailListArrayList.get(position).getSend_date();

                    long diffTime = diffTime(prev_date_string, this_date_string);
                    Log.e("krm0219", diffTime + " " + prev_date_string + "  " + this_date_string + "  " + messageDetailListArrayList.get(position).getMessage());

                    if (1 <= diffTime) {

                        text_message_detail_receiver_id.setVisibility(View.VISIBLE);
                    } else {

                        text_message_detail_receiver_id.setVisibility(View.GONE);
                    }
                }
            }

            if (Build.VERSION.SDK_INT >= 24) {

                text_message_detail_receive_message.setText(Html.fromHtml(item.getMessage(), Html.FROM_HTML_MODE_LEGACY));
            } else {

                text_message_detail_receive_message.setText(Html.fromHtml(item.getMessage()));
            }
        } else if (item.getAlign().equalsIgnoreCase("RIGHT")) {
            /* from driver > to customer */
            layout_message_detail_receive_message.setVisibility(View.GONE);
            layout_message_detail_send_message.setVisibility(View.VISIBLE);

            text_message_detail_send_date.setText(item.getSend_date());

            if (Build.VERSION.SDK_INT >= 24) {

                text_message_detail_send_message.setText(Html.fromHtml(item.getMessage(), Html.FROM_HTML_MODE_LEGACY));
            } else {

                text_message_detail_send_message.setText(Html.fromHtml(item.getMessage()));
            }
        }


        //  연속으로 메세지 보낸거 확인 > layout param 바꾸기 / Date  visible/gone
        if (position < messageDetailListArrayList.size() - 1) {

            String this_align = messageDetailListArrayList.get(position).getAlign();
            String next_align = messageDetailListArrayList.get(position + 1).getAlign();

            String this_date_string = messageDetailListArrayList.get(position).getSend_date();
            String next_date_string = messageDetailListArrayList.get(position + 1).getSend_date();

            long diffTime = diffTime(this_date_string, next_date_string);

            if (next_align.equalsIgnoreCase(this_align)) {  // left > left  // right > right

                if (1 <= diffTime) {

                    layout_message_detail.setPadding(dpTopx(15), dpTopx(5), dpTopx(15), dpTopx(10));
                    text_message_detail_receive_date.setVisibility(View.VISIBLE);
                    text_message_detail_send_date.setVisibility(View.VISIBLE);
                } else {        // 1분 미만으로 동일한 사람이 입력!

                    layout_message_detail.setPadding(dpTopx(15), dpTopx(5), dpTopx(15), dpTopx(5));
                    text_message_detail_receive_date.setVisibility(View.GONE);
                    text_message_detail_send_date.setVisibility(View.GONE);
                }
            } else {        // left > right  // right > left
                layout_message_detail.setPadding(dpTopx(15), dpTopx(5), dpTopx(15), dpTopx(10));
                text_message_detail_receive_date.setVisibility(View.VISIBLE);
                text_message_detail_send_date.setVisibility(View.VISIBLE);
            }
        } else {        // last

            layout_message_detail.setPadding(dpTopx(15), dpTopx(5), dpTopx(15), dpTopx(10));
            text_message_detail_receive_date.setVisibility(View.VISIBLE);
            text_message_detail_send_date.setVisibility(View.VISIBLE);
        }

        return view;
    }

    private int dpTopx(float dp) {

        int pixel = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, mContext.getResources().getDisplayMetrics());
        return pixel;
    }

    private long diffTime(String old_date_string, String date_string) {

        long diff_time = 100;

        try {

            DateFormat dateFormat = null;
            if(calledFragment.equalsIgnoreCase("C")) {

                dateFormat = new SimpleDateFormat("yyyy-MM-dd a HH:mm");
            } else if(calledFragment.equalsIgnoreCase("A")) {

                dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm");
            }

            Date old_date = dateFormat.parse(old_date_string);
            Date date = dateFormat.parse(date_string);

            long old_date_time = old_date.getTime();
            long date_time = date.getTime();

            diff_time = (date_time - old_date_time) / 60000;

        } catch (Exception e) {

            e.printStackTrace();
            diff_time = 100;
        }

        return diff_time;
    }
}

File: mylibrary/src/main/java/com/giosis/library/message/MessageDetailAdapter.kt
package com.giosis.library.message

import android.annotation.SuppressLint
import android.content.Context
import android.os.Build
import android.text.Html
import android.util.Log
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.giosis.library.databinding.ItemMessageDetailBinding
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*

class MessageDetailAdapter(
    var context: Context,
    var items: ArrayList<MessageDetailResult>?,
    var calledFragment: String
) : RecyclerView.Adapter<MessageDetailAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding =
            ItemMessageDetailBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    class ViewHolder(val binding: ItemMessageDetailBinding) : RecyclerView.ViewHolder(binding.root)

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {

        val item = items!![position]

        if (item.align.equals("LEFT", ignoreCase = true)) {
            /* from customer, admin > to driver */

            holder.binding.layoutReceiveMessage.visibility = View.VISIBLE
            holder.binding.layoutSendMessage.visibility = View.GONE

            if (item.sender_id == "") {
                holder.binding.textReceiverId.text = "null"
            } else {
                holder.binding.textReceiverId.text = item.sender_id
            }
            holder.binding.textReceiveDate.text = item.send_date


            // Customer(Admin) ID   visible/gone
            if (position == 0) {
                holder.binding.textReceiverId.visibility = View.VISIBLE
            } else {  // 1 이상일 때, 나의 이전  ALIGN  비교

                val prevAlign = items!![position - 1].align
                val thisAlign = items!![position].align

                if (thisAlign != prevAlign) {   // right > left

                    holder.binding.textReceiverId.visibility = View.VISIBLE
                } else {

                    val prevDate = items!![position - 1].send_date
                    val thisDate = items!![position].send_date
                    val diffTime = diffTime(prevDate, thisDate)
                    Log.e(
                        "Message",
                        diffTime.toString() + " " + prevDate + "  " + thisDate + "  " + items!![position].message
                    )

                    if (1 <= diffTime) {
                        holder.binding.textReceiverId.visibility = View.VISIBLE
                    } else {
                        holder.binding.textReceiverId.visibility = View.GONE
                    }
                }
            }

            if (Build.VERSION.SDK_INT >= 24) {
                holder.binding.textReceiveMessage.text =
                    Html.fromHtml(item.message, Html.FROM_HTML_MODE_LEGACY)
            } else {
                holder.binding.textReceiveMessage.text = Html.fromHtml(item.message)
            }
        } else if (item.align.equals("RIGHT", ignoreCase = true)) {
            /* from driver > to customer */

            holder.binding.layoutReceiveMessage.visibility = View.GONE
            holder.binding.layoutSendMessage.visibility = View.VISIBLE
            holder.binding.textSendDate.text = item.send_date

            if (Build.VERSION.SDK_INT >= 24) {
                holder.binding.textSendMessage.text =
                    Html.fromHtml(item.message, Html.FROM_HTML_MODE_LEGACY)
            } else {
                holder.binding.textSendMessage.text = Html.fromHtml(item.message)
            }
        }


        //  연속으로 메세지 보낸거 확인 > layout param 바꾸기 / Date  visible/gone
        if (position < items!!.size - 1) {

            val thisAlign = items!![position].align
            val nextAlign = items!![position + 1].align
            val thisDate = items!![position].send_date
            val nextDate = items!![position + 1].send_date
            val diffTime = diffTime(thisDate, nextDate)

            if (nextAlign.equals(thisAlign, ignoreCase = true)) {  // left > left  // right > right
                if (1 <= diffTime) {

                    holder.binding.layoutMessageDetail.setPadding(
                        dpTopx(15f),
                        dpTopx(5f),
                        dpTopx(15f),
                        dpTopx(10f)
                    )
                    holder.binding.textReceiveDate.visibility = View.VISIBLE
                    holder.binding.textSendDate.visibility = View.VISIBLE
                } else {        // 1분 미만으로 동일한 사람이 입력!

                    holder.binding.layoutMessageDetail.setPadding(
                        dpTopx(15f),
                        dpTopx(5f),
                        dpTopx(15f),
                        dpTopx(5f)
                    )
                    holder.binding.textReceiveDate.visibility = View.GONE
                    holder.binding.textSendDate.visibility = View.GONE
                }
            } else {        // left > right  // right > left

                holder.binding.layoutMessageDetail.setPadding(
                    dpTopx(15f),
                    dpTopx(5f),
                    dpTopx(15f),
                    dpTopx(10f)
                )
                holder.binding.textReceiveDate.visibility = View.VISIBLE
                holder.binding.textSendDate.visibility = View.VISIBLE
            }
        } else {        // last

            holder.binding.layoutMessageDetail.setPadding(
                dpTopx(15f),
                dpTopx(5f),
                dpTopx(15f),
                dpTopx(10f)
            )
            holder.binding.textReceiveDate.visibility = View.VISIBLE
            holder.binding.textSendDate.visibility = View.VISIBLE
        }
    }

    override fun getItemCount(): Int {

        if (items != null) {

            if (0 < items!!.size) {
                return items!!.size
            }
        }

        return 0
    }


    private fun dpTopx(dp: Float): Int {
        return TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP,
            dp,
            context.resources.displayMetrics
        ).toInt()
    }

    @SuppressLint("SimpleDateFormat")
    private fun diffTime(old_date_string: String, date_string: String): Long {

        var diffTime: Long

        try {
            var dateFormat: DateFormat? = null

            if (calledFragment.equals("C", ignoreCase = true)) {

                dateFormat = SimpleDateFormat("yyyy-MM-dd a HH:mm")
            } else if (calledFragment.equals("A", ignoreCase = true)) {

                dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm")
            }

            val oldDate = dateFormat!!.parse(old_date_string)
            val date = dateFormat.parse(date_string)
            val oldDateTime = oldDate.time
            val dateTime = date.time

            diffTime = (dateTime - oldDateTime) / 60000
        } catch (e: Exception) {

            e.printStackTrace()
            diffTime = 100
        }

        return diffTime
    }
}